server/src/main/java/org/elasticsearch/transport/TransportHandshaker.java

```diff
@@ -149,24 +160,75 @@ public String executor() {
         }
     }
 
-    static final class VersionHandshakeResponse extends TransportResponse {
+    static final class HandshakeRequest extends TransportRequest {
 
         private final Version version;
 
-        VersionHandshakeResponse(Version version) {
+        HandshakeRequest(Version version) {
             this.version = version;
         }
 
-        private VersionHandshakeResponse(StreamInput in) throws IOException {
+        HandshakeRequest(StreamInput streamInput) throws IOException {
+            super(streamInput);
+            BytesReference remainingMessage;
+            try {
+                remainingMessage = streamInput.readBytesReference();
+            } catch (EOFException e) {
+                remainingMessage = null;
+            }
+            if (remainingMessage == null) {
+                version = null;
+            } else {
+                try (StreamInput messageStreamInput = remainingMessage.streamInput()) {
+                    this.version = Version.readVersion(messageStreamInput);
+                }
+            }
+        }
+
+        @Override
+        public void readFrom(StreamInput in) {
+            throw new UnsupportedOperationException("usage of Streamable is to be replaced by Writeable");
+        }
+
+        @Override
+        public void writeTo(StreamOutput streamOutput) throws IOException {
+            super.writeTo(streamOutput);
+            assert version != null;
+            try (BytesStreamOutput messageStreamOutput = new BytesStreamOutput(4)) {
+                Version.writeVersion(version, messageStreamOutput);
+                BytesReference reference = messageStreamOutput.bytes();
+                streamOutput.writeBytesReference(reference);
+            }
+        }
+    }
+
+    static final class HandshakeResponse extends TransportResponse {
+
+        private final Version responseVersion;
```

Comment:
If the plan is to replace `internal:transport/handshake` with this message, does it also need the `ClusterName`? Today we rely on the fact that the transport handshake validates that the cluster names match during discovery - I think that we don't otherwise verify that we're talking to nodes from the right cluster.

Reply:
This is follow-up work. We need the version the the request to bootstrap the follow-up work. Once we have the version in the request, we can serialize arbitrary responses that are compatible with the remote node's version.